Add a new 'ps' variant to `Duration` for picoseconds.

kevinhartman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Our `delay` and `box` instructions support picoseconds, so it's a gap that we can't use `Duration` to represent these. 

### Details and comments
The new variant requires support from QPY ~and thus bumps its version~. QASM3 does not have native support for picoseconds, so we convert to nanoseconds during export if they are encountered (this is what we currently do for the old-style duration and unit in a delay).

mtreinish commented on qiskit/qasm3/exporter.py:
```diff
@@ -1336,6 +1336,9 @@ def visit_value(self, node, /):
                 return ast.DurationLiteral(node.value.value(), ast.DurationUnit.SAMPLE)
+            if unit == "ps":
+                # QASM doesn't natively support picoseconds.
+                return ast.DurationLiteral(1000 * node.value.value(), ast.DurationUnit.NANOSECOND)
```
Is this the right direction? Isn't 1 picosecond .001 nanoseconds? So shouldn't we be dividing here instead of multiplying? Right now if it's 1 ps this will export 1000 ns (or 1 microsecond).

kevinhartman replied:
Yeah, you're correct. I actually copied this from [here](https://github.com/Qiskit/qiskit/blob/ff9a3b65eca7592db2e72f3e68bd5227385092bc/qiskit/qasm3/exporter.py#L1180), which means we have an existing bug too.

I opened up this PR to fix the original bug:

- #14895

mtreinish commented on test/python/qasm3/test_export.py:
```diff
@@ -750,6 +753,7 @@ def test_delay_statement(self):
                 "delay[2000ns] qr[1];",
                 "delay[s / 2.0] qr[1];",
                 "delay[s * (1000dt / 200.0ns) + t] qr[0];",
+                "delay[s * (1000dt / 2000.0ns) + t] qr[0];",
```
I don't think this is the correct expected result, 2ps is .002ns, 2000ns is 2 us. I left another comment in the exporter where the arithmetic is incorrect for the conversion.

kevinhartman replied:
Fixed in e0b012ddffc5171b94884e1039e1bb9bdb37a806

mtreinish reviewed (commented):
Overall this is pretty straightforward, there's just one issue inline for the qasm converter.

kevinhartman committed:
Add a new 'ps' variant to Duration for picoseconds.

The new variant requires support from our QPY version. QASM3 does
not have native support for picoseconds, so we convert to nanoseconds
during export if they are encountered.

kevinhartman committed:
Address review comments.

mtreinish commented on crates/circuit/src/duration.rs:
```diff
@@ -70,14 +72,17 @@ impl Duration {
 
     fn __repr__(&self) -> String {
         match self {
+            Duration::ps(t) => format!("Duration::ps({t})"),
```
Should this be?
```suggestion
            Duration::ps(t) => format!("Duration.ps({t})"),
```
Since this is the Python repr, also all the others use `.`.

kevinhartman replied:
Fixed in 914fb73a76079d7efe5722516687b956d5381fbd

mtreinish commented on test/python/circuit/test_circuit_load_from_qpy.py:
```diff
@@ -2250,6 +2250,22 @@ def test_pre_v14_rejects_stretch_expr(self, version):
             dump(qc, fptr, version=version)
 
+    @ddt.idata(range(14, 16))
+    def test_pre_v16_rejects_ps_duration(self, version):
```
Is there a test that validates we can roundtrip a picosecond unit delay in version 16?

kevinhartman replied:
Good point. I added it to our `test_qpy_roundtrip_control_flow_expr` test circuit. Which it actually turns out was completely dead code (looks like a copy-paste mistake). It's wired up now, and luckily everything works 😅.

17fa4d914b7bd22a032bd386d5b331a7a269c605

kevinhartman committed:
Fix repr for Duration.ps.

kevinhartman committed:
Fix roundtrip transpile expr test and add Duration.ps.

mtreinish reviewed (approved):
This LGTM now, thanks for the updates and adding this.
